breakpad-symbols/src/http.rs

```diff
@@ -142,46 +151,70 @@ fn file_key(module: &(dyn Module + Sync), file_kind: FileKind) -> FileKey {
-    // instance's partially written results.
-    //
-    // tempfile is designed explicitly for this purpose, and will handle all
-    // the platform-specific details and do its best to cleanup if things
-    // crash.
-
+async fn create_cache_file(final_path: &Path) -> io::Result<PartialDownload> {
     // First ensure that the target directory in the cache exists
     let base = final_path.parent().ok_or_else(|| {
         io::Error::new(
             io::ErrorKind::Other,
             format!("Bad cache path: {:?}", final_path),
         )
     })?;
-    fs::create_dir_all(&base)?;
-
-    NamedTempFile::new_in(tmp_path)
+    fs::create_dir_all(&base).await?;
+
+    // Try to create a new file in the final destination with `.part` added to the extension.
+    // Use an attempt counter to vaguely allow multiple processes to race on this, without them
+    // having to cooperate in any way. If there's only one instance this will just work reasonably.
+    let dest_path = final_path.to_owned();
+    let mut temp_path = dest_path.clone();
+    let mut new_ext = temp_path.extension().unwrap().to_owned();
+    new_ext.push(".part");
+    let mut num_attempts = 0;
+    loop {
+        // Try successive .0, .1, ... extensions to not fall over for garbage and other processes
+        let mut cur_ext = OsString::new();
+        cur_ext.push(&new_ext);
+        cur_ext.push(".");
+        cur_ext.push(num_attempts.to_string());
+        temp_path.set_extension(&cur_ext);
+
+        match fs::File::create(&temp_path).await {
+            Ok(file) => {
+                return Ok(PartialDownload {
+                    file,
+                    dest_path,
+                    temp_path,
+                });
+            }
+            Err(e) => {
+                if let io::ErrorKind::AlreadyExists = e.kind() {
+                    if num_attempts > 5 {
+                        // Give up, something fucky is happening
+                        // (this isn't great, but, WIP?)
+                        return Err(e);
+                    } else {
+                        num_attempts += 1;
+                        continue;
+                    }
+                }
+                return Err(e);
+            }
+        }
+    }
 }
 
-fn commit_cache_file(mut temp: NamedTempFile, final_path: &Path, url: &Url) -> io::Result<()> {
+async fn commit_cache_file(mut download: PartialDownload, url: &Url) -> io::Result<()> {
     // Append any extra metadata we also want to be cached as "INFO" lines,
     // because this is an established format that parsers will ignore the
     // contents of by default.
 
     // INFO URL allows us to properly report the url we retrieved a symbol file
     // from, even when the file is loaded from our on-disk cache.
     let cache_metadata = format!("INFO URL {}\n", url);
-    temp.write_all(cache_metadata.as_bytes())?;
-
-    // TODO: don't do this
-    if final_path.exists() {
-        fs::remove_file(final_path)?;
-    }
```

Comment:
I genuinely forget why this exists. I have a feeling I was corrupting files and then all the caching machinery assumed the files weren't corrupt, so I just wanted existing files to get blown away..? But that doesn't really make sense because at this point we already discovered the file wasn't there... This is a super weird path.